sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala

```diff
@@ -73,7 +74,8 @@ private[sql] class SessionState(
     resourceLoaderBuilder: () => SessionResourceLoader,
     createQueryExecution: LogicalPlan => QueryExecution,
     createClone: (SparkSession, SessionState) => SessionState,
-    val columnarRules: Seq[ColumnarRule]) {
+    val columnarRules: Seq[ColumnarRule],
+    val queryStagePrepRules: Seq[Rule[SparkPlan]]) {
```

Comment:
SessionState is a critical concept/class in Spark SQL. Adding `queryStagePrepRules` into SessionState looks weird to me. WDYT?

cc @hvanhovell

Reply:
Preferably both `columnarRules` and `queryStagePrepRules` should be passed directly to the thing that is using them. This is not entirely easy because of how we deal with post-planning rules in QueryExecution. It would be great if someone could move those into the builder.

Reply:
It's very hard. The extension rules are injected in the `BaseSessionStateBuilder`, and we need to carry it in `SessionState` to propagate it further. The custom `columnarRules` and `queryStagePrepRules` are no different from custom analyzer/optimizer/planner rules.

Reply:
It is just refactoring right (famous last words)? You pull out the rules from query execution, put them in the session state builder, and manipulate them there.

Reply:
Maybe I missed something. The session state builder is used to build `SessionState`. `SparkSession` or `QueryExecition` doesn't hold session state builder instance anymore.

Are you suggesting we create something like `Analyzer`, `Optimizer`, etc. to wrap the post-planning rules in `QueryExecution`, so that we just pass that wrapper around?